nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py

```diff
@@ -0,0 +1,653 @@
+        in the MegatronGPT class.
+        """
+        # Get embeddings for text tokens and insert virtual token embeddings
+        if inference:
+            input_embeds = self.embed_input_inference(input_ids, taskname_ids)
+        else:
+            input_embeds = self.embed_input_train(input_ids, taskname_ids)
+
+        position_embeddings = self.model.model.language_model.embedding.position_embeddings(position_ids)
+        encoder_input = input_embeds + position_embeddings
+
+        # Call forward on GPT model with preprocessed embeddings
+        if self.float_type == torch.float32:
+            output = self.model.model(
+                input_ids=None,
+                position_ids=None,
+                encoder_input=encoder_input,
+                attention_mask=attention_mask,
+                labels=labels,
+                set_inference_key_value_memory=set_inference_key_value_memory,
+                inference_max_sequence_len=inference_max_sequence_len,
+            )
+        else:
+            with torch.autocast(device_type="cuda", dtype=self.float_type):
+                output = self.model.model(
+                    input_ids=None,
+                    position_ids=None,
+                    encoder_input=encoder_input,
+                    attention_mask=attention_mask,
+                    labels=labels,
+                    set_inference_key_value_memory=set_inference_key_value_memory,
+                    inference_max_sequence_len=inference_max_sequence_len,
+                )
+
+        return output
+
+    def embed_input_train(self, input_ids: Tensor, taskname_ids: Tensor):
+        """
+        Replaces the virtual tokens in the input_ids with embeddings 
+        calculated from either the 'prompt_table' or 'prompt_encoder'. 
+        The virtual token placeholders have token_ids listed in
+        `self.pseudo_token_ids`.
+
+        params:
+            input_ids: the input token ids
+            taskname_ids: the NLP task tag token ids
+        returns:
+            the token embedding for the LM model.
+        """
+        # Replace virtual token ids with padding for forward pass through vocab embeddings
+        discrete_token_ids = input_ids.clone()
+        discrete_token_ids[(input_ids >= self.pseudo_token_ids_start)] = self.pad_token_id
+        discrete_token_embeds = self.word_embeddings(discrete_token_ids).clone()
+
+        # Find the indicies where virtual tokens should be inserted
+        virtual_token_locations = input_ids >= self.pseudo_token_ids_start
+
+        # If there are no virtual tokens, just return discrete token embeds
+        if not virtual_token_locations.any():
+            return discrete_token_embeds
+
+        # Get virtual token embeddings from the prompt table or prompt encoder
+        if self.virtual_prompt_source == 'prompt-table':
+            virtual_token_embeddings = [self.prompt_table(task_id_num) for task_id_num in taskname_ids]
```

Comment:
Oh, good point. I'll try it using torch.tensor operations.

Reply:
One solution is to design the prompt table into a high dimensional tensor.
something like `[task_dim, virtual_token_dim, embedding_dim]`
then you need to encode different task-name/ids into number from [0, task_dim-1]. 
Good thing about this approach is you can look up a list of tasks easily by one tensor operation (gather method or tensor slices). 
Bad thing about this approach is it assumes each task uses same amount of virtual tokens. It may cause some waste.
I will let you to decide it is a good idea or not.

Reply:
I think I'm going to keep the prompt table as is to allow for different amount of virtual tokens per task during inference. I think any bottle neck using this hash table approach might cause is solved in cases where we really want the model to be most performant since the FasterTransformer team will be concatenating virtual token embeddings with the vocab embeddings.

Reply:
I changed the code a bit and changed the CI test to make sure inference works with virtual prompts that are different lengths.